sycl/plugins/esimd_emulator/pi_esimd_emulator.hpp

```diff
@@ -104,48 +110,96 @@ struct _pi_mem : _pi_object {
 
   pi_context Context;
 
+  // To be used for piEnqueueMemBufferMap
   char *MapHostPtr = nullptr;
 
-  // Mutex for load/store accessing
+  // Mutex for load/store accessing to CM-managed surfaces from memory
+  // instrinsic implementations
   std::mutex mutexLock;
 
-  // Surface index used by CM
-  int SurfaceIndex;
+  // Surface index
+
+  unsigned int SurfaceIndex;
+  // Supplementary data to keep track of the mappings of this memory
+  // created with piEnqueueMemBufferMap
+  struct Mapping {
+    // The offset in the buffer giving the start of the mapped region.
+    size_t Offset;
+    // The size of the mapped region.
+    size_t Size;
+  };
+
+  // The key is the host pointer representing an active mapping.
+  // The value is the information needed to maintain/undo the mapping.
+  // TODO : std::unordered_map is imported from L0.
+  // Use std::stack for strict LIFO behavior checking?
+  std::unordered_map<void *, Mapping> Mappings;
+  // Supporing multi-threaded mapping/unmapping calls
+  std::mutex MappingsMutex;
 
   virtual ~_pi_mem() = default;
```

Comment:
The destructor is virtual, however base class (_pi_object) destructor is not.

Reply:
This is imported from level_zero. Will update `_pi_object` here.

Reply:
Done.

Reply:
Couple of notes:
- Those added virtual destructors don't do anything and unlikely will, so this is just extra lines of code to maintain. I don't think they should've been added. If _pi_buffer needs to execute some additional code in destructor, that's when `virtual ~pi_buffer::pi_buffer()` should be added, but not as `=default`
- there is no need in default destructor in _pi_mem as well

Reply:
Per Konst's comment, I'll remove all `virtual` destructors.

Reply:
Yes, as mentioned above it just needs to be consistent. Virtual or not - depends on usage model: virtual destructor is needed if you will delete the object using pointer to a base class